sdk/nodejs/resource.ts

```diff
@@ -29,6 +29,52 @@ export abstract class Resource {
      // tslint:disable-next-line:variable-name
      /* @internal */ private readonly __pulumiResource: boolean = true;
 
+    /**
+     * The optional parent of this resource.
+     */
+    // tslint:disable-next-line:variable-name
+    /* @internal */ public readonly __parentResource: Resource | undefined;
+
+    /**
+     * The child resources of this resource.  We use these (only from a ComponentResource) to allow
+     * code to dependOn a ComponentResource and have that effectively mean that it is depending on
+     * all the CustomResource children of that component.
+     *
+     * Important!  We only walk through ComponentResources.  They're the only resources that serve
+     * as an aggregation of other primitive (i.e. custom) resources.  While a custom resource can be
+     * a parent of other resources, we don't want to ever depend on those child resource.  If we do,
+     * it's simple to end up in a situation where we end up depending on a child resource that has a
+     * data cycle dependency due to the data passed into it.
+     *
+     * An example of how this would be bad is:
+     *
+     * ```ts
+     *     var c1 = new CustomResource("c1");
+     *     var c2 = new CustomResource("c2", { parentId: c1.id }, { parent: c1 });
+     *     var c3 = new CustomResource("c3", { parentId: c1.id }, { parent: c1 });
+     * ```
+     *
+     * The problem here is that 'c2' has a data dependency on 'c1'.  If it tries to wait on 'c1' it
+     * will walk to the children and wait on them.  This will mean it will wait on 'c3'.  But 'c3'
+     * will be waiting in the same manner on 'c2', and a cycle forms.
+     *
+     * This cannot happen with ComponentResources as they do not have any data flowing into them.
```

Comment:
This isn't strictly true: components can still participate in inputs/outputs from `registerResource` just like custom resources. However, that scenario is so obscure that I think that it's safe to assume that it will never happen.